src/Microsoft.AspNetCore.Routing/DispatcherMiddleware.cs

```diff
@@ -0,0 +1,114 @@
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Matchers;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    internal sealed class DispatcherMiddleware
+    {
+        private readonly MatcherFactory _matcherFactory;
+        private readonly ILogger _logger;
+        private readonly IOptions<DispatcherOptions> _options;
+        private readonly RequestDelegate _next;
+
+        private Task<Matcher> _initializationTask;
+
+        public DispatcherMiddleware(
+            MatcherFactory matcherFactory,
+            IOptions<DispatcherOptions> options,
+            ILogger<DispatcherMiddleware> logger,
+            RequestDelegate next)
+        {
+            if (matcherFactory == null)
+            {
+                throw new ArgumentNullException(nameof(matcherFactory));
+            }
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            if (next == null)
+            {
+                throw new ArgumentNullException(nameof(next));
+            }
+
+            _matcherFactory = matcherFactory;
+            _options = options;
+            _logger = logger;
+            _next = next;
+        }
+
+        public async Task Invoke(HttpContext httpContext)
+        {
+            var feature = new EndpointFeature();
+            httpContext.Features.Set<IEndpointFeature>(feature);
+
+            // There's an inherit race condition between waiting for init and accessing the matcher
+            // this is OK because once `_matcher` is initialized, it will not be set to null again.
+            var matcher = await InitializeAsync();
+
+            await matcher.MatchAsync(httpContext, feature);
+            if (feature.Endpoint != null)
+            {
+                Log.EndpointMatched(_logger, feature);
+            }
+
+            await _next(httpContext);
+        }
+
+        // Initialization is async to avoid blocking threads while reflection and things
+        // of that nature take place.
+        //
+        // We've seen cases where startup is very slow if we  allow multiple threads to race 
+        // while initializing the set of endpoints/routes. Doing CPU intensive work is a 
+        // blocking operation if you have a low core count and enough work to do.
+        private Task<Matcher> InitializeAsync()
+        {
+            if (_initializationTask != null)
+            {
+                return _initializationTask;
+            }
+
+            var initializationTask = new TaskCompletionSource<Matcher>();
+            if (Interlocked.CompareExchange<Task<Matcher>>(
+                ref _initializationTask,
+                initializationTask.Task,
+                null) == null)
```

Comment:
You feel smarter doing it this way.

Reply:
Serioustalk: I think of these as different patterns, and so the idea of using a lock didn't even enter my brain. 

This is a latch, it's either initialized or not, but once it's initialized, you never need to think about it again. When I see something like this in code I understand that's how it will behave without having to read much more of the context.

When I see a lock, I assume that something more complicated than a contended initialization is happening.

Reply:
Ok. Compared to locking I guess the disadvantage of this way is allocating additional TCS while waiting for initialization to complete.
Advantage is you don't need a flag to say initialization is complete and a lock object.

Reply:
I don't feel strongly about the details, this seemed idiomatic to me.